File: exercise/day19/src/test/java/blog/ArticleTests.java
package blog;

import org.assertj.core.api.ThrowingConsumer;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.function.Function;

import static blog.ArticleBuilder.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.instancio.Instancio.create;

class ArticleTests {
    private Article article;

    @Test
    void should_add_comment_in_an_article() {
        when(article -> article.addComment(COMMENT_TEXT, AUTHOR));
        then(article -> {
            assertThat(article.getComments()).hasSize(1);
            assertComment(article.getComments().get(0), COMMENT_TEXT, AUTHOR);
        });
    }

    @Test
    void should_add_comment_in_an_article_containing_already_a_comment() {
        final var newComment = create(String.class);
        final var newAuthor = create(String.class);

        when(ArticleBuilder::commented, article -> article.addComment(newComment, newAuthor));
        then(article -> {
            assertThat(article.getComments()).hasSize(2);
            assertComment(article.getComments().last(), newComment, newAuthor);
        });
    }

    private static void assertComment(Comment comment, String commentText, String author) {
        assertThat(comment.text()).isEqualTo(commentText);
        assertThat(comment.author()).isEqualTo(author);
    }

    private void when(ArticleBuilder articleBuilder, Function<Article, Article> act) throws CommentAlreadyExistException {
        article = act.apply(
                articleBuilder.build()
        );
    }

    private void when(Function<Article, Article> act) {
        when(anArticle(), act);
    }

    private void when(Function<ArticleBuilder, ArticleBuilder> options, Function<Article, Article> act) {
        when(options.apply(anArticle()), act);
    }

    private void then(ThrowingConsumer<Article> act) {
        act.accept(article);
    }

    @Nested
    class Fail {
        @Test
        void when__adding_an_existing_comment() {
            var article = anArticle()
                    .commented()
                    .build();

            assertThatThrownBy(() -> {
                article.addComment(article.getComments().get(0).text(), article.getComments().get(0).author());
            }).isInstanceOf(CommentAlreadyExistException.class);
        }
    }
}

File: exercise/c#/Day19/Day19.Tests/ArticleTests.cs
using FluentAssertions;
using Xunit;
using static Day19.Tests.ArticleBuilder;

namespace Day19.Tests
{
    public class ArticleTests
    {
        private Article _article;
        private readonly Bogus.Randomizer _random = new();

        [Fact]
        public void Should_Add_Comment_In_An_Article()
        {
            Given(AnArticle());
            When(article => article.AddComment(CommentText, Author));
            Then(article =>
            {
                article.Comments.Should().HaveCount(1);
                AssertComment(article.Comments[0], CommentText, Author);
            });
        }

        [Fact]
        public void Should_Add_Comment_In_An_Article_Containing_Already_A_Comment()
        {
            var newComment = _random.String(10);
            var newAuthor = _random.String(3);

            Given(AnArticle().Commented());
            When(article => article.AddComment(newComment, newAuthor));
            Then(article =>
            {
                article.Comments.Should().HaveCount(2);
                AssertComment(article.Comments[1], newComment, newAuthor);
            });
        }

        private static void AssertComment(Comment comment, string expectedComment, string expectedAuthor)
        {
            comment.Text.Should().Be(expectedComment);
            comment.Author.Should().Be(expectedAuthor);
            comment.CreationDate.Should().Be(DateOnly.FromDateTime(DateTime.Now));
        }

        public class Fail
        {
            [Fact]
            public void When_Adding_An_Existing_Comment()
            {
                var article = AnArticle().Build()
                    .AddComment(CommentText, Author);

                var act = () => article.AddComment(CommentText, Author);
                act.Should().Throw<CommentAlreadyExistException>();
            }
        }

        private void Given(ArticleBuilder articleBuilder) => _article = articleBuilder.Build();
        private void When(Func<Article, Article> act) => _article = act(_article);
        private void Then(Action<Article> act) => act(_article);
    }
}